homeassistant/components/zamg/sensor.py

```diff
@@ -199,187 +212,97 @@ class ZamgSensorEntityDescription(SensorEntityDescription, ZamgRequiredKeysMixin
 )
 
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
-    add_entities: AddEntitiesCallback,
+    async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the ZAMG sensor platform."""
-    name = config[CONF_NAME]
+
+    probe = ZamgData(session=async_get_clientsession(hass))
+
     latitude = config.get(CONF_LATITUDE, hass.config.latitude)
     longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
-
-    station_id = config.get(CONF_STATION_ID) or closest_station(
-        latitude, longitude, hass.config.config_dir
-    )
-    if station_id not in _get_ogd_stations():
-        _LOGGER.error(
-            "Configured ZAMG %s (%s) is not a known station",
-            CONF_STATION_ID,
+    station_id = config.get(CONF_STATION_ID)
+    if station_id not in await probe.zamg_stations():
+        LOGGER.warning(
+            "Configured station_id %s could not be found at zamg, adding the nearest weather station instead",
             station_id,
         )
-        return
+        station_id = await probe.closest_station(latitude, longitude)
+    probe.set_default_station(station_id)
+
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2023.1.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
 
-    probe = ZamgData(station_id=station_id)
-    try:
-        probe.update()
-    except (ValueError, TypeError) as err:
-        _LOGGER.error("Received error from ZAMG: %s", err)
-        return
+    # No config entry exists and configuration.yaml config exists, trigger the import flow.
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.data.get(CONF_STATION_ID):
+            return
```

Comment:
I think this should be moved up higher, before the queries and creating the issue. Otherwise every single configured sensor or weather entity in yaml would fire a request and create an issue.

Same for the weather platform.

To avoid this duplicate code (here and in `weather.py`) you could just call the import step here and do `self._abort_if_unique_id_configured` in the ConfigFlow to abort if the station_id was already configured. Also the probe thing and the issue creation can be done there to avoid calling that multiple times.

Reply:
For a entry with a ```station_id``` it's possible to check for existence and abort the ConfigFlow, but a entry in the ```configuration.yaml``` without ```station_id``` or just coordinates will always trigger a new request to get the right ```station_id```.

Yes, avoiding duplicate code is always good. I Will dig into it.

Reply:
Oh, I guess its not a big deal to have some extra requests for the few months the migration period lasts 😉

Reply:
I reworked the code and put the common stuff into the import flow which is now also reducing the calls to zamg (one time refresh of all possible stations).
The other thing about creating a issue: Every time the issue is gone after HA restarts (even with `is_persistent=True`), that's why i call it every time.